### ISSUE TITLE
Floating point literals are not parsed accurately

### ISSUE THREAD

**Issue by born2defy** [purescript/purescript#2115]
I am getting some strange results with Number equality.  This is with psc 0.8.5.  See below:

```
> let x =  roundTo 2  0.166  
> x
> 0.17
> x == 0.17
> false
```

**Comment by hdgarrood** [comment]
Can you please show the definition of `roundTo`? I can't find it on pursuit.

**Comment by hdgarrood** [comment]
Ah, seems like an issue with the `Show Number` instance. I can reproduce what is probably the same problem with just `Prelude`:
```
> 0.17000000000000001
0.17
> 0.17000000000000001 == 0.17
false
```

**Comment by born2defy** [comment]
Sure - it is taken from the mozilla page on rounding here : https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/round

The javascript is here:

```
exports.decimalAdjustF = function decimalAdjustF(type, value, exp) {
    // If the exp is undefined or zero...
    if (typeof exp === 'undefined' || +exp === 0) {
      return Math[type](value);
    }
    value = +value;
    exp = +exp;

    // Shift
    value = value.toString().split('e');
    value = Math[type](+(value[0] + 'e' + (value[1] ? (+value[1] - exp) : -exp)));
    // Shift back
    value = value.toString().split('e');
    return +(value[0] + 'e' + (value[1] ? (+value[1] + exp) : exp));
  }
```

imported into PS like this:

```
foreign import decimalAdjustF :: Fn3 String Number Int Number

decimalAdjust :: String -> Number -> Int -> Number
decimalAdjust = runFn3 decimalAdjustF

roundTo :: Int -> Number -> Number
roundTo x = flip (decimalAdjust "round") (fix x)

-- | Make sure the value passed to decimalAdjust are negative
fix :: Int -> Int
fix x = if x < 0 then x else negate x

```

**Comment by hdgarrood** [comment]
Ok, I've worked it out. Turns out the definition of `roundTo` doesn't matter, it's because PS and JS disagree about rounding of literals. Perhaps we should copy whatever JS does?

```purescript
module Main where

import Prelude
import Control.Monad.Eff.Console (print)

main = do
  let x = 0.17000000000000001
  let y = 0.17
  print x
  print y
  print (x == y)
```

Produces the following output:

```
0.17
0.16999999999999998
false
```

The generated JS is:

```javascript
// Generated by psc version 0.8.5.0
"use strict";
var Prelude = require("../Prelude");
var Control_Monad_Eff_Console = require("../Control.Monad.Eff.Console");
var Control_Monad_Eff = require("../Control.Monad.Eff");
var main = function __do() {
    Control_Monad_Eff_Console.print(Prelude.showNumber)(0.17)();
    Control_Monad_Eff_Console.print(Prelude.showNumber)(0.16999999999999998)();
    return Control_Monad_Eff_Console.print(Prelude.showBoolean)(0.17 === 0.16999999999999998)();
};
module.exports = {
    main: main
};
```

**Comment by born2defy** [comment]
Wow - very surprised to see the value being generated for `y` in that instance.

**Comment by AppShipIt** [comment]
This sounds a little like #2093 again? I think comparing small fractions is the common gotcha but technically you'd hit similar issues with `17000000000000001===17000000000000001//=>false`. Although `y` is surprising I have no idea how to account for precision in the comparison.

**Comment by hdgarrood** [comment]
JS doesn't use any concept of precision when comparing numbers; they have to be the same value to be considered equal, see: http://www.ecma-international.org/ecma-262/6.0/#sec-samevalue. PureScript uses the same equality for `Number` values as JS does, and I am fairly confident that this is the best choice: using some delta for comparisons would mean we'd lose transitivity, and also I don't think it's what anyone would expect. I don't think this is all that much like #2093, I don't think it has anything to do with Number equality really.

The problem, as far as I can tell, is that if you have a `Number` value, and you call `show` on it to turn it into a string, and then copy that into a PureScript program and use it as a literal, you can end up with a different value to what you started with. This should not happen. Floating point has gotchas, but they're not severe enough that we can't preserve this property.

Whether the bug is in the `Show Number` instance or the PureScript parser depends on your point of view, but either way they should be brought into line with each other.

Something weird must be happening inside the compiler to make the literal `0.17000000000000001` come out as a literal `0.17` in the JS, and likewise to make `0.17` come out as `0.16999999999999998`; it seems to me that this is the bit that wants changing. The `Show Number` instance just uses `Number.prototype.toString` (and adds a ".0" if necessary) and that seems right.

**Comment by AppShipIt** [comment]
> The Show Number instance just uses Number.prototype.toString (and adds a ".0" if necessary) and that seems right.

Would calling `#toPrecision(16)` (precision picked by messing in the console but could probably be calculated) for `Show Number` be more predictable?

**Comment by jplatte** [comment]
> The `Show Number` instance just uses `Number.prototype.toString`

The two give me different results on node 6.1.0 and psci 0.8.5.0:

```
% pulp psci
 ____                 ____            _       _   
|  _ \ _   _ _ __ ___/ ___|  ___ _ __(_)_ __ | |_ 
| |_) | | | | '__/ _ \___ \ / __| '__| | '_ \| __|
|  __/| |_| | | |  __/___) | (__| |  | | |_) | |_ 
|_|    \__,_|_|  \___|____/ \___|_|  |_| .__/ \__|
                                       |_|        

:? shows help
> let x = 1e25
> x
9.999999999999999e+24

---

% node
> var x = 1e25
undefined
> x.toString()
'1e+25'
```

**Comment by hdgarrood** [comment]
@jplatte yes, I expect something similar is happening here. The problem, I think, is that we can't trust literals in purescript. What happens if you write `let x = 1e25` in PS, compile it, and look at the produced JS?

**Comment by hdgarrood** [comment]
Just checked and `let x = 1e25` in PS comes out as `9.999999999999999e24` in the JS, so it is the compiler again, not the `Show Number` instance.

@AppShipIt perhaps, but I am fairly sure that calling `Number.prototype.toString` without arguments will already be calculating the minimum necessary precision required to display the number value unambiguously, so we might as well reuse that, I think.

**Comment by jplatte** [comment]
Yeah, right, that's the problem. My first guess at the cause of this problem would have been float to double conversion, but apparently the margin of error is a lot bigger there:

```
λ> import GHC.Float
λ> float2Double 0.17
0.17000000178813934
λ> float2Double 1e25
9.999999562023526e24
```

**Comment by born2defy** [comment]
For the moment I am solving this particular case (comparing a calculated value to a literal) by rounding both to the same decimal point which appears to preserve equality.  i.e.

```
> let x = roundTo 2  0.166
> x
> 0.17
> let y = 0.17
> x == y 
> false
> x == roundTo 2 y
> true
```

**Comment by hdgarrood** [comment]
It looks like it is parsec's fault:

```
$ stack ghci
> :load Language.PureScript.Parser.Lexer
> let parseNumber = P.try (PT.float tokenParser)
> P.runParser parseNumber () "<stdin>" "0.17"
Right 0.16999999999999998
```

**Comment by hdgarrood** [comment]
I think this is https://github.com/aslatter/parsec/issues/35.

**Comment by born2defy** [comment]
If so, judging from the issue thread it doesn't appear that it will be fixed any time soon (at least in the original library).

**Comment by hdgarrood** [comment]
No, but it's easy to fix locally.

**Comment by hdgarrood** [comment]
This will be fixed in 0.9. Thanks for the report!
